### DISCUSSION TITLE
Select device with Mac OS

### DISCUSSION THREAD

**Discussion by Badsheep** [mikebrady/shairport-sync#1411] [Q&A] [answered]
Hello,

I just installed Shairport-sync and it seems to work nicely with Mac OS.

Just a question : I'd like to route the sound in a particular device (an aggregate device called "ZIC") and I don't konw where to set it. I tried to had the line 
dev = "ZIC";
in the shairport-sync.conf file, but no success, the sound still outputs in the audio device defined as default in Mac OS.

I tried to launch the app with some arguments :  
shairport-sync ao-o dev=ZIC
But it's the same.

According to the [Libao documentation](https://xiph.org/ao/doc/drivers.html), the "dev" argument should be usable but I don't understand how...

An alternative would be to be able to route in differents channels (3,4 instead of 1,2 for example), but according to the Libao documentation channel mapping is not possible in Mac OS so I don't think it may work this way.

Thanks for any help !

**Comment by Badsheep** [comment]
I found a [closed thread ](https://github.com/mikebrady/shairport-sync/issues/718) about it and it seems like it’s not possible to choose an audio device, shaiport-sync just outputs to the default soundcard. So I think I’ll have to make a script to select the sound device needed, then launch shairport-sync, then select another audio device.
If anyone knows how to avoid this, please let me know ! Thanks

**Comment by dge8** [comment, marked as answer]
To anyone who stumbles across this in future, I've recently been down this rabbit hole and AFAICT the `dev` argument to `libao` has been broken since 2016 (it was introduced in 2013, so it's been broken longer than it ever worked!). There's a 2yo pending pull request [here](https://gitlab.xiph.org/xiph/libao/-/merge_requests/7) that fixes this.

Until the pull request is merged, you can manually install the patched version like so (assuming homebrew on macOS):
1. Install `shairport-sync` with `brew install shairport-sync`
2. Remove `libao` 1.2.2 with `brew remove libao --ignore-dependencies`
3. Copy the following into a file called `libao.rb` (in your home directory, or elsewhere):
```
class Libao < Formula
  desc "Cross-platform Audio Library"
  homepage "https://www.xiph.org/ao/"
  url "https://gitlab.xiph.org/mackyle/libao/-/archive/e0d7c43117b0566fbc7c8f7fd3e3c5d4f24c0fbb/libao-e0d7c43117b0566fbc7c8f7fd3e3c5d4f24c0fbb.tar.gz", using: :homebrew_curl
  sha256 "35b077d98b7403cc84406ba720d718d6a5d812ffd1189c7e675bceb384f12825"
  license "GPL-2.0"
  version "1.2.2-p1"
  head "https://gitlab.xiph.org/mackyle/libao.git", branch: "patched"

  depends_on "autoconf" => :build
  depends_on "automake" => :build
  depends_on "libtool" => :build
  depends_on "pkg-config" => :build

  def install
    ENV["AUTOMAKE_FLAGS"] = "--include-deps"
    system "./autogen.sh"

    args = %W[
      --disable-dependency-tracking
      --prefix=#{prefix}
      --enable-static
    ]

    system "./configure", *args
    system "make", "install"
  end

  test do
    (testpath/"test.cpp").write <<~EOS
      #include <ao/ao.h>
      int main() {
        ao_initialize();
        return 0;
      }
    EOS
    system ENV.cc, "test.cpp", "-I#{include}", "-L#{lib}", "-lao", "-o", "test"
    system "./test"
  end
end
```
4. Navigate to the formula you just created and install the patched `libao` with `brew install --build-from-source ./libao.rb` 
5. Run `shairport-sync` using the command line `shairport-sync -o ao -- -n DeviceName`, or `shairport-sync -o ao -- -o dev=DeviceName`. `DeviceName` can be the exact device name as it appears in Audio MIDI Setup, or any (case-insensitive) substring, so long as it only matches one device.

**Comment by Badsheep** [comment]
Hi dge8, I just tried what you suggest and it works like a charm !
Thank you very much !
